### DISCUSSION TITLE
Multipartition for a time interval in a dagster framework

### DISCUSSION THREAD

**Discussion by lemonade-thinking** [dagster-io/dagster#17993] [Q&A] [answered]
Hi all,

I came across a question on defining and using static partition with multi partition recently and I am not sure how to resolve these questions.

Essentially, I would like to pair the start and stop timestamp in a partition to initiate and stop a series of defined assets (i.e. this series of assets should only be run within the ``start_timestamp_partition`` and ``stop_timestamp_partition``. I am using a static partition for this reason because the timestamps do not fit into the hourly, weekly or monthly patterns.

For example, this means that ``2019-02-03 07:00`` should only be paired with ``2019-03-19 15:15:00``, ``2019-04-20 23:55:00`` should only be paired with ``2019-07-02 10:20:00`` ... for each run.  From what I understand, I think that multipartition seems to be doing cross -pairing with each element in the list, resulting in 16 pairs altogether, instead of 4 pairs only. 

How should I best proceed with implementing this partition in dagster? 

```from dagster import asset, Definitions, StaticPartitionsDefinition, MultiPartitionsDefinition

start_timestamp_partition = StaticPartitionsDefinition([
    '2019-02-03 07:00',
    '2019-04-20 23:55:00',
    '2019-07-02 11:20:00',
    '2019-09-22 15:05:00'])

stop_timestamp_partition = StaticPartitionsDefinition([
    '2019-03-19 15:15:00',
    '2019-07-02 10:20:00',
    '2019-09-22 08:30:00',
    '2019-12-31 23:55'])


missing_gap_partition = MultiPartitionsDefinition(
{
    "start_asset_series": start_timestamp_partition,
    "stop_asset_series": stop_timestamp_partition
}
)

@asset(partitions_def=missing_gap_partition)
def multi_partition_asset(context):
    partition = context.partition_key.keys_by_dimension
    context.log.info(
        f"start: {partition['start_asset_series']}\n" +
        f"stop: {partition['stop_asset_series']}"
    )
)`

**Comment by sryza** [comment, marked as answer]
> I think that multipartition seems to be doing cross -pairing with each element in the list

That's correct. `MultiPartitionsDefinition` takes the cross-product of the sets of partitions in the included `PartitionsDefinition`s.

For this case, I would recommend just using a single `StaticPartitionsDefinition` where each partition represents a pair of timestamps. E.g.

```python
my_time_interval_partitions = StaticPartitionsDefinition([
    '2019-02-03 07:00,2019-04-20 23:55:00',
    '2019-07-02 11:20:00,2019-09-22 15:05:00',
])
```

**Reply by lemonade-thinking** [reply]
Hi @sryza, thank you for your helpful feedback.

I have tried implementing this static partition solution and integrating within an asset, but was not able to get the graph running:

![image](https://github.com/dagster-io/dagster/assets/101148138/9d7d8e34-baf7-43e8-96b8-b747eac5d595)

The error message that I obtained is:

> dagster._core.errors.DagsterExecutionHandleOutputError: Error occurred while handling output "result" of step "test_partition"

I have two questions about this implementation:

1. How to integrate this partition into an asset? Did I make any mistakes in the above snippets that cause the errors?
2. I am not sure what is the Python type of ``StaticPartitionsDefinition`` output? Is the type a ``list`` or a ``string``? Is the above implementation correct to query the date in the asset for further analysis?

Thank you for your suggestions!

**Reply by sryza** [reply]
Are you able to include a full stack trace? That error normally wraps a more specific error.

**Reply by lemonade-thinking** [reply]
Hi @sryza: thank you for your reply. Here is the full stack trace:

```
dagster._core.errors.DagsterExecutionStepExecutionError: Error occurred while executing op "test_partition":
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\execute_plan.py", line 275, in dagster_event_sequence_for_step
    for step_event in check.generator(step_events):
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\execute_step.py", line 478, in core_dagster_event_sequence_for_step
    for user_event in _step_output_error_checked_user_event_sequence(
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\execute_step.py", line 160, in _step_output_error_checked_user_event_sequence
    for user_event in user_event_sequence:
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\execute_step.py", line 95, in _process_asset_results_to_events
    for user_event in user_event_sequence:
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\compute.py", line 212, in execute_core_compute
    for step_output in _yield_compute_results(step_context, inputs, compute_fn, compute_context):
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\compute.py", line 181, in _yield_compute_results
    for event in iterate_with_context(
  File "myvenv\lib\site-packages\dagster\_utils\__init__.py", line 463, in iterate_with_context
    with context_fn():
  File "mypath\contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\utils.py", line 84, in op_execution_error_boundary
    raise error_cls(
The above exception was caused by the following exception:
IndexError: list index out of range
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\utils.py", line 54, in op_execution_error_boundary
    yield
  File "myvenv\lib\site-packages\dagster\_utils\__init__.py", line 465, in iterate_with_context
    next_output = next(iterator)
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\compute_generator.py", line 131, in _coerce_op_compute_fn_to_iterator
    result = invoke_compute_fn(
  File "myvenv\lib\site-packages\dagster\_core\execution\plan\compute_generator.py", line 125, in invoke_compute_fn
    return fn(context, **args_to_pass) if context_arg_provided else fn(**args_to_pass)
  File "mypath\test.py", line 12, in test_partition
    end_date = partition_interval[1].split(",")[1]
```

**Reply by sryza** [reply]
It looks like the base of the error is coming from the code you posted, not from inside Dagster. I'd recommend printing out `partition_interval` and making sure that it has the structure you expect.
